tests/freshness.rs

```diff
@@ -182,6 +189,264 @@ fn changing_features_is_ok() {
     assert_that(p.cargo("build"),
                 execs().with_status(0)
                        .with_stdout(""));
+
+    assert_that(p.cargo("build").arg("--features").arg("foo"),
+                execs().with_status(0)
+                       .with_stderr("\
+[..]Compiling foo v0.0.1 ([..])
+[FINISHED] debug [unoptimized + debuginfo] target(s) in [..]
+"));
+}
+
+#[test]
+fn changing_profiles_caches_targets() {
+    let p = project("foo")
+        .file("Cargo.toml", r#"
+            [package]
+            name = "foo"
+            authors = []
+            version = "0.0.1"
+
+            [profile.dev]
+            panic = "abort"
+
+            [profile.test]
+            panic = "unwind"
+        "#)
+        .file("src/lib.rs", "");
+
+    assert_that(p.cargo_process("build"),
+                execs().with_status(0)
+                       .with_stderr("\
+[..]Compiling foo v0.0.1 ([..])
+[FINISHED] debug [unoptimized + debuginfo] target(s) in [..]
+"));
+
+    assert_that(p.cargo("test"),
+                execs().with_status(0)
+                       .with_stderr("\
+[..]Compiling foo v0.0.1 ([..])
+[FINISHED] debug [unoptimized + debuginfo] target(s) in [..]
+[RUNNING] target[..]debug[..]deps[..]foo-[..][EXE]
+[DOCTEST] foo
+"));
+
+    /* Targets should be cached from the first build
+       XXX Sadly these cannot be cached since the "symlink" step is
+           not separate from the "compile" step. Packages which link
+           to top level binaries eg (deps/foo-abc123 -> foo) are forced
+           to do an extra recompile here.
+    */
```

Comment:
Both parts of this test requires symlink. Currently, this test is passing because it's queueing a recompile (which is doing the relink), but with your symlink-in-fresh idea, I was able to get this to work with no recompiles! Hooray.